main.tf

```diff
@@ -1,64 +1,100 @@
-# ------------------------------------------------------------------------------------------------
+# -------------------------------------------------------------------------------------------------
 
+# Create roles
 resource "aws_iam_role" "roles" {
-  count = "${var.role_count}"
+  for_each = { for role in var.roles : role.name => role }
 
-  name        = "${lookup(var.roles[count.index], "name")}"
-  path        = "${lookup(var.roles[count.index], "path", "") == "" ? var.role_path : lookup(var.roles[count.index], "path")}"
-  description = "${lookup(var.roles[count.index], "desc", "") == "" ? var.role_desc : lookup(var.roles[count.index], "desc")}"
+  name        = lookup(each.value, "name")
+  path        = lookup(each.value, "path", "") == "" ? var.role_path : lookup(each.value, "path")
+  description = lookup(each.value, "desc", "") == "" ? var.role_desc : lookup(each.value, "desc")
 
   # This policy defines who/what is allowed to use the current role
-  assume_role_policy = "${file(lookup(var.roles[count.index], "trust_policy_file"))}"
+  assume_role_policy = file(lookup(each.value, "trust_policy_file"))
 
   # The boundary defines the maximum allowed permissions which cannot exceed.
   # Even if the policy has higher permission, the boundary sets the final limit
-  permissions_boundary = "${lookup(var.roles[count.index], "permissions_boundary", "")}"
+  permissions_boundary = lookup(var.permissions_boundaries, each.key, "")
 
   # Allow session for X seconds
-  max_session_duration  = "${var.max_session_duration}"
-  force_detach_policies = "${var.force_detach_policies}"
+  max_session_duration  = var.max_session_duration
+  force_detach_policies = var.force_detach_policies
 
-  tags = "${merge(
-    map("Name", lookup(var.roles[count.index], "name")),
+  tags = merge(
+    map("Name", lookup(each.value, "name")),
     var.tags
-  )}"
+  )
+
 }
 
-# ------------------------------------------------------------------------------------------------
-# Create Policies for above Roles
-# ------------------------------------------------------------------------------------------------
 
-resource "aws_iam_policy" "policies" {
-  count = "${var.role_count}"
+# -------------------------------------------------------------------------------------------------
+# Attach Policies to Role
+# -------------------------------------------------------------------------------------------------
+
+# Attach customer managed policies
+resource "aws_iam_role_policy_attachment" "policy_attachments" {
+  for_each = local.role_policies
 
-  name        = "${lookup(var.roles[count.index], "policy_name")}"
-  path        = "${lookup(var.roles[count.index], "policy_path", "") == "" ? var.policy_path : lookup(var.roles[count.index], "policy_path")}"
-  description = "${lookup(var.roles[count.index], "policy_desc", "") == "" ? var.policy_desc : lookup(var.roles[count.index], "policy_desc")}"
+  role       = replace(each.key, format(":%s", each.value.name), "")
```

Comment:
If you add the role_name as regular attribute of the map you don't have to do this replacing.

Reply:
The key of each of the map lists is now consistent with `<role-name>:<other-part>`. Imho, this would be too much of a change for all the local loops again.